### PR TITLE
Update NATS.Client.Hosting package as NATS.Extensions.Microsoft.DependencyInjection

### LINKED ISSUE #392: Update NATS.Client.Hosting package as NATS.Extensions.Microsoft.DependencyInjection
### Proposed change

Update NATS.Client.Hosting package as NATS.Extensions.Microsoft.DependencyInjection and mark NATS.Client.Hosting obsolete.

### Use case

Confusing for new users on NuGet search results. Also x.Extensions.Microsoft.DependencyInjection seems to be the convention.

cc @mrpmorris thanks for the suggestion

### LINKED ISSUE #417: Simplify JSON serialization setup
### Proposed change

Currently because we're AOT compatible context or ad-hoc JSON serialization requires a bit of setup to register in the connection:

```csharp
var jsonRegistry = new NatsJsonContextSerializerRegistry(MyJsonContext.Default);
return opts with { SerializerRegistry = jsonRegistry };
```

We should be able to simplify this when using hosting extensions.

### PR DESCRIPTION
**Pull request by rickdotnet**
Aims to resolve #392 and resolve #417 

The builder pattern is fairly common and easy to pick up. This also leaves it fairly open to future extension.

```cs
var services = new ServiceCollection();

services.AddNats(builder => builder.AddJsonSerialization(MyJsonContext.Default));
services.AddNats(builder => builder.WithPoolSize(2).WithKey(key2));
```

### CODE REVIEW

**Comment by rickdotnet** [comment]
As is, this would break the existing service registration when switching to the new library. Which, wouldn't be the end of the world, but breaking for anyone passing arguments today. It's an easy swap to the new registration pattern. We could change either signature as well and bring the old extension over. However, as stands, they both overlap when no arguments are passed. So, if we want to maintain compatibility, we could require an argument in the new method and they won't overlap.

File: src/NATS.Client.Hosting/NatsHostingExtensions.cs [hunk 1]
```diff
@@ -15,6 +15,7 @@ public static class NatsHostingExtensions
     [System.Diagnostics.CodeAnalysis.SuppressMessage("StyleCop.CSharp.SpacingRules", "SA1009:Closing parenthesis should not be preceded by a space", Justification = "Required for conditional build.")]
     [System.Diagnostics.CodeAnalysis.SuppressMessage("StyleCop.CSharp.ReadabilityRules", "SA1111:Closing parenthesis should be on the same line as the last parameter", Justification = "Required for conditional build.")]
     [System.Diagnostics.CodeAnalysis.SuppressMessage("StyleCop.CSharp.ReadabilityRules", "SA1113:Comma should be on the same line as previous parameter", Justification = "Required for conditional build.")]
+    [Obsolete("This package is obsolete. Use NATS.Extensions.Microsoft.DependencyInjection instead.")]
```
[the remark this answers is not in the document]
Comment (mtmk) [reply src/NATS.Client.Hosting/NatsHostingExtensions.cs:18]: maybe as you suggested we could leave deprecating the old one for another PR

File: src/NATS.Extensions.Microsoft.DependencyInjection/NatsHostingExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.DependencyInjection.Extensions;
+using Microsoft.Extensions.Logging;
+using NATS.Client.Core;
+
+namespace NATS.Extensions.Microsoft.DependencyInjection;
+
+public static class NatsHostingExtensions
+{
+    public static IServiceCollection AddNats(this IServiceCollection services, Action<NatsBuilder>? buildAction = null)
```
[the remark this answers is not in the document]
Comment (mtmk) [reply src/NATS.Extensions.Microsoft.DependencyInjection/NatsHostingExtensions.cs:7]: I might have a solution. In [Aspire impl it's called AddNatsClient()](https://github.com/dotnet/aspire/blob/main/src/Components/Aspire.NATS.Net/AspireNatsClientExtensions.cs#L33). We can keep the names the same. if the name is different we won't have compatibility issues.

File: tests/NATS.Extensions.Microsoft.DependencyInjection.Tests/MyJsonContext.cs [hunk 3]
```diff
@@ -0,0 +1,8 @@
+using System.Text.Json.Serialization;
+
+namespace NATS.Extensions.Microsoft.DependencyInjection.Tests;
+
+[JsonSerializable(typeof(MyData))]
+internal partial class MyJsonContext : JsonSerializerContext;
+
+public record MyData(string Name);
```
Comment (mtmk) [inline tests/NATS.Extensions.Microsoft.DependencyInjection.Tests/MyJsonContext.cs:8]: create a property for this with getter+setter. net6 have problems with source generators with init properties.

File: src/NATS.Client.Hosting/NatsHostingExtensions.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (rickdotnet) [reply src/NATS.Client.Hosting/NatsHostingExtensions.cs]: Removed this.

File: tests/NATS.Extensions.Microsoft.DependencyInjection.Tests/MyJsonContext.cs (hunk 3 above)
Comment (rickdotnet) [reply tests/NATS.Extensions.Microsoft.DependencyInjection.Tests/MyJsonContext.cs]: Done

File: src/NATS.Extensions.Microsoft.DependencyInjection/NatsHostingExtensions.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (rickdotnet) [reply src/NATS.Extensions.Microsoft.DependencyInjection/NatsHostingExtensions.cs]: Named it `AddNatsClient`, which, I think is what you meant by this.

**Comment by mtmk** [comment]
thanks @rickdotnet

**Pull request merged**